Add support for markdown > alerts via markdig

kzu opened the pull request:
Partial fix for #10125

Markdig supports alert blocks as documented at https://github.com/xoofx/markdig/blob/master/src/Markdig.Tests/Specs/AlertBlockSpecs.md.

Bumping markdig and adding `.UseAlertBlocks()` now renders the expected html.

Not sure how to go about adding the styles as needed.

kzu committed:
Add support for markdown > alerts via markdig

Markdig supports alert blocks as documented at https://github.com/xoofx/markdig/blob/master/src/Markdig.Tests/Specs/AlertBlockSpecs.md.

Bumping markdig and adding `.UseAlertBlocks()` now renders the expected html.

Fixes #10125

erdembayar commented on tests/NuGetGallery.Facts/Services/MarkdownServiceFacts.cs:
```diff
@@ -105,6 +105,18 @@ public void EncodesHtmlInMarkdownWithAdaptiveHeader(string originalMd, string ex
             [InlineData("![](http://www.otherurl.net/fake.jpg)", "<p><img src=\"https://www.otherurl.net/fake.jpg\" alt=\"\" /></p>", true, false)]
             [InlineData("![](http://www.otherurl.net/fake.jpg)", "<p><img src=\"https://www.otherurl.net/fake.jpg\" class=\"img-fluid\" alt=\"alternate text is missing from this package README image\" /></p>", true, true)]
             [InlineData("## License\n\tLicensed under the Apache License, Version 2.0 (the \"License\");", "<h3 id=\"license\">License</h3>\n<pre><code>Licensed under the Apache License, Version 2.0 (the &quot;License&quot;);\n</code></pre>", false, true)]
+            [InlineData(
```
Do you have screenshot for how it would look like?

kzu replied:
nope, since I can't get the site running, so this is the best I can do. I uploaded the initial style too, just guessing at what the right styles should be though...

erdembayar replied:
By the way unit test is failing.

erdembayar replied:
Have you checked this https://github.com/NuGet/NuGetGallery?tab=readme-ov-file#run-the-gallery-website-locally for building the code on your local? If there's problem with running on your local please let us know and we can work out.

kzu replied:
build.ps1 shows a failure in:

```
  sign-binaries failed with 1 error(s) (0.0s)
    ...\NuGetGallery\build\sign-binaries.proj(24,5): error MSB3202: The project file ""...\NuGetGallery\src\**\*.csproj"" was not found.
```

But otherwise, things seem to run just fine.

kzu committed:
Make sure alerts use default font color

If we don't unset it, the default bootstrap p style turns
dark mode illegible (uses white-ish color). This seems
to be an issue specific to the very old version of Bootstrap
being used (v3.x).

When running the site locally, I could not find any other
.css (or .less) files being built or linked from the page,
so I placed the only needed style in the only other style
file other than bootstrap.

kzu commented:
I guess as a test PR, it showcases that for someone that can actually run the site, it should be fairly straightforward to add support for this :)

lyndaidaii commented:
> I guess as a test PR, it showcases that for someone that can actually run the site, it should be fairly straightforward to add support for this :)

Thank you for your contribution. I could help to test it.

erdembayar commented on src/NuGetGallery/Content/Site.css:
```diff
@@ -1374,20 +1374,20 @@ ul.accordion {
-        ul.accordion li.accordion-item .accordion-expand-link:hover {
-            text-decoration: underline;
-        }
+            ul.accordion li.accordion-item .accordion-expand-link:hover {
```
Please remove unnecessary space in front.

jimmylewis replied:
@erdembayar This is a common CSS indentation pattern when one selector extends the preceding rule selector.  It helps visualize that this selector is both related and more specific. (FWIW in VS, this is enabled by default as the Hierarchical Indentation setting under CSS options.)

erdembayar commented:
> > I guess as a test PR, it showcases that for someone that can actually run the site, it should be fairly straightforward to add support for this :)
> 
> Thank you for your contribution. I could help to test it.

Thanks, Lynn. Please add screenshot here.

erdembayar commented:
@lyndaidaii 
Have you had chance to test this? If not then I can do it, please let me know.

erdembayar commented on src/NuGetGallery/Content/Site.css:
```diff
@@ -1374,20 +1374,20 @@ ul.accordion {
 
         ul.accordion li.accordion-item .accordion-expand-link {
-            background: none!important;
+            background: none !important;
```
Also this change should have been done in less file rather than this file?

kzu replied:
I have no idea where things are in your code base. Feel free to close this issue which was mostly a demonstration that fixing the issue is mostly adding one line of code and whichever styles you want for the site 🙏

kzu replied:
would you mind pointing at *which* `.less` file you're referring to?

![image](https://github.com/user-attachments/assets/9ab22485-1d85-472a-ae5f-a1bcf97a4a0e)

erdembayar replied:
@NuGet/gallery-team 

Do you know which less file we need to make changes to? I never change any of them.

kzu commented:
Dark mode: 

![image](https://github.com/user-attachments/assets/0d944ee4-1767-4459-8ad0-1ff090b20006)

Light mode;

![image](https://github.com/user-attachments/assets/32922911-8489-4e81-9a13-acbc5681b4f4)

kzu commented:
any updates @erdembayar @lyndaidaii ?

erdembayar commented:
> any updates @erdembayar @lyndaidaii ?

Sorry, we'll review this week or at the latest next week. We have other priorities at the moment and need some time.

lyndaidaii commented:
> @lyndaidaii Have you had chance to test this? If not then I can do it, please let me know.

I had trouble in building gallery in local since we merged a couple of repos into one.  Haven't checked in since then. If you could build gallery in local fine, could you help take a look at UI?

kzu commented:
Any updates?

erdembayar commented:
> Any updates?

As you can see, unit tests are passing, which is good news. However, I couldn't test the actual rendering view part because our Gallery app has a binding redirect issue due to a recent package update. We need to fix it first. In the meantime, could you please upload the NuGet package with the readme you demoed to nuget.org or share the Readme.md file with me here? It would make my testing easier.
![image](https://github.com/user-attachments/assets/4e0511d2-33b3-4fc1-a5b4-c1508d74514f)

erdembayar commented:
Please rebase your branch using our latest branch. 
It seems rendering is not really workin.

![image](https://github.com/user-attachments/assets/d41e065b-ef24-4189-9dd1-857ff9bd53ba)

I used below Readme.md
```

## About This Project

Hello this is readme file.

[!NOTE]
This is a note 2024
```

lyndaidaii commented:
Looks like is missing some of css file.

kzu commented:
@erdembayar the sample package I used is available as source (pretty bare, mind you) at https://github.com/kzu/Observable.

It's basically this readme: [readme.md](https://github.com/user-attachments/files/17285775/readme.md)

And a nuspec like

```xml
<?xml version="1.0" encoding="utf-8"?>
<package xmlns="http://schemas.microsoft.com/packaging/2012/06/nuspec.xsd">
  <metadata>
    <id>Observable</id>
    <version>0.2.0</version>
    <authors>Daniel Cazzulino</authors>
    <requireLicenseAcceptance>false</requireLicenseAcceptance>
    <license type="expression">MIT</license>
    <licenseUrl>https://licenses.nuget.org/MIT</licenseUrl>
    <projectUrl>https://github.com/devlooped/Observable</projectUrl>
    <description>Sample.</description>
    <readme>readme.md</readme>
    <tags>dotnet csharp</tags>
  </metadata>
  <files>
    <file src="readme.md" target="readme.md" />
  </files>
</package>
```
[Observable.0.2.0.nupkg.zip](https://github.com/user-attachments/files/17285808/Observable.0.2.0.nupkg.zip)


which I ran nuget pack on.

kzu commented:
I rebased, and I'm getting proper styles:

![image](https://github.com/user-attachments/assets/caf330fa-7cb4-4a72-88ff-522a381dd1a6)

The app isn't requesting anything other than the fabric.css and bootstrap.min.css: 

![image](https://github.com/user-attachments/assets/d45a5f73-e172-4af1-868e-780dcec4c325)

And those seem to be the default alert styles in bootstrap indeed:

https://getbootstrap.com/docs/3.4/components/#alerts

That's for the right version of bootstrap in use by the gallery: 

![image](https://github.com/user-attachments/assets/79ae6be9-2b7a-49d7-9e90-2e017948fc3f)

erdembayar commented:
> I rebased, and I'm getting proper styles:

Did you push your rebased code? I don't see it. 
![image](https://github.com/user-attachments/assets/d2319a56-cef9-4fb8-baed-53820cb6020e)

kzu commented:
@erdembayar what I meant is that the rebase didn't make any difference. Just in case, I just pushed anyways after a new rebase.

The NOTE will never look like that because there's no such [built-in style for that in bootstrap](https://getbootstrap.com/docs/3.4/components/#alerts). I have no idea where you  put styles that aren't bootstrap, but if you tell me where, I can add one for NOTE since extending the alert styles is pretty easy in bootstrap. But I think this is going too far. I'm not a designer :). 

The PR showcases how adding support for this is straightforward. The unit test shows that the proper classes are emitted. Now all that's left is for you to bring that note style that's used elsewhere in MS docs sites and put it somewhere you like. I didn't find any place with custom styles, which is why I just put something in the `fabric.css`, but I've no idea if that's the "right" place.

kzu commented:
BTW, I'm happy to just copy over from github css the style for the above, but I've no idea if that's what you want design-wise for nuget.org :)

lyndaidaii commented:
@kzu, thank you for your contribution. Great work. No worries at all. I will work on the style.

erdembayar commented:
> BTW, I'm happy to just copy over from github css the style for the above, but I've no idea if that's what you want design-wise for nuget.org :)

Thank you. It looks @lyndaidaii  is going to take care of `Note`.

erdembayar commented:
> @kzu, thank you for your contribution. Great work. No worries at all. I will work on the style.

For your reference, I found actual Note sample used in our documentation which is also markdown format.
See bottom of https://learn.microsoft.com/en-us/nuget/reference/errors-and-warnings/nu1605, also as you can see coloring is `Tip` little bit different too.

![image](https://github.com/user-attachments/assets/4ce89ee9-789b-4b3e-bdcb-7ecb8ef20f72)

kzu commented:
Yeah, that looks better IMO

Commit:
update css style

lyndaidaii commented:
Follow the css from bootstrap, since we have been using bootstrap for previous alerts.  reference: https://getbootstrap.com/docs/4.0/components/alerts/
![image](https://github.com/user-attachments/assets/5f8db6ac-b29a-4886-8c46-b9c5f253bd0a)

lyndaidaii commented:
After adding CSS: 
![image](https://github.com/user-attachments/assets/24e7ae95-48b6-46fe-808e-39cc6c2db98c)
![image](https://github.com/user-attachments/assets/21b88aee-93c1-4057-bc67-cd354363998d)

lyndaidaii commented on tests/NuGetGallery.Facts/Services/MarkdownServiceFacts.cs:
```diff
@@ -105,12 +105,23 @@ public void EncodesHtmlInMarkdownWithAdaptiveHeader(string originalMd, string ex
+                <p class="mb-0">This is a note</p>
+                </div>
+                """,
+                false, true)]
```
there is no way we could test for css style.

erdembayar replied:
But could you still add test for text?

lyndaidaii replied:
added unit tests.

Commit:
add unit test

erdembayar reviewed (dismissed):
@lyndaidaii  
Thanks for driving this to a conclusion! 👏  
Maybe we can to make coloring more stronger for `Alert, Tip` etc, not blocking, we can do it in another PR. 
Please wait for a while to let other team members review.

zivkan commented on src/NuGetGallery/Services/MarkdownService.cs:
```diff
@@ -208,6 +208,7 @@ private RenderedMarkdownResult GetHtmlFromMarkdownMarkdig(string markdownString,
                 .UseTaskLists()
                 .UseEmojiAndSmiley()
                 .UseAutoLinks()
+                .UseAlertBlocks()
```
I think it's important for the markdown dialect to use with NuGet readmes should be designed with intent, rather than ad-hoc as customers want features.

NuGet.Client is in the process of adding readme support to Visual Studio's Package Manager UI:
* https://github.com/NuGet/NuGet.Client/pull/6110

If nuget.org supports markdown features that VS does not, that won't be good for customers, either package consumers or package authors.

cc @jgonz120

kzu replied:
How does that align with the readme support already there in GitHub? The alert format and markup is fairly standard at this point, I'd say...

zivkan replied:
from: https://github.github.com/gfm/

> GitHub Flavored Markdown, often shortened as GFM, is the dialect of Markdown that is currently supported for user content on GitHub.com and GitHub Enterprise.

so, they admit it's a specific dialect, not some kind of reference, common, or minimal set that all markdown renders are expected to implement.

> How does that align with the readme support already there in GitHub?

GitHub's markdown dialect also supports things like:

* [collapsed sections](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/organizing-information-with-collapsed-sections)
* [mermaid js diagrams](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-diagrams)
* [mathematical expressions](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/writing-mathematical-expressions)

If the suggestion is that NuGet should support GFM, then these are the things that have to work as well, as well as an expectation that every time GitHub add a new markdown feature, both nuget.org and NuGet client are expected to implement it as well. I don't think these features are valuable enough to warrant the complexity to have to re-implement them in both nuget.org and VS.

So, GitHub supporting something in their dialect isn't justification to me that NuGet should support it

What concerns me is that customers will expect the readme to render the same in VS as on nuget.org.  If NuGet package readmes don't render "correctly" on github.com, that would be unfortunate, but it's less important to me because it's not a primary location for viewing package readmes, and not all packages use github.

If the desire is to use a single readme as both the package readme as well as the repo readme, that's understandable, so perhaps NuGet's readme dialect could be defined as a subset of GFM.  Or perhaps we can encourage package authors to make the repo readme focused on contribution instructions, how to build and debug the repo locally, stuff that's not relevant to the package readme. It's still valuable for the repo readme to have a short introduction to what the package is and how to use it, but it can then have a link to nuget.org to see the full package readme, which can have more detailed information.

> The alert format and markup is fairly standard at this point, I'd say...

That's fine. Then the NuGet server team can talk to the client team, get agreement that it's feasible to support the same formatting in both.  I'm just saying the server team shouldn't extend the readme markdown dialect without talking to the client team.

I'm not saying don't do it. But I believe that customers are going to get a bad experience if a package author who doesn't use VS, or doesn't test their readme in VS, writes a readme that doesn't render correctly in VS. I'm sure that customers who notice the difference will report bugs to us. What should our response be in such a case?  I don't want extra support costs just because a small number of package authors of open source packages on github, want to use the readme to render the same on github and nuget.org.

Therefore, I believe the readme should be considered part of the "protocol" that should have both client and server teams agree to, not one team adding it without even talking to the other team.

kzu replied:
> If the suggestion is that NuGet should support GFM, then these are the things that have to work as well, as well as an expectation that every time GitHub add a new markdown feature, both nuget.org and NuGet client are expected to implement it as well.

Not at all. But alerts are a very common documentation aid and it would be entirely unproductive to make up a brand new format, or worse, continue to NOT support it at all when it's ubiquitous even in MS learn site.

> So, GitHub supporting something in their dialect isn't justification to me that NuGet should support it

The main reason would be familiarity for OSS authors, IMHO.

> What concerns me is that customers will expect the readme to render the same in VS as on nuget.org.

What concerns me is that you may come up with an entirely incompatible way just for the purposes of your "nuget markdown" and force authors to learn yet another flavor rather than a subset of what they already know.

> If the desire is to use a single readme as both the package readme as well as the repo readme, that's understandable, so perhaps NuGet's readme dialect could be defined as a subset of GFM. 

Makes sense to be. The supported URLs in links are already one such case.

> > The alert format and markup is fairly standard at this point, I'd say...
> 
> That's fine. 

That was the point of this PR. To showcase that it's trivial to add support for it on the gallery. You could just render plain blockquotes in VS without styling in v1 for readmes of that simplifies things. I doubt anyone will have high expectations for that readme visualization since it's a non existent feature at the moment.

jgonz120 replied:
The MarkdownPreview component we're using doesn't currently allow for the alerts. I can check with the team that owns the component to see the feasibility of adding the functionality.

erdembayar commented:
We need to have cross team discussion between client and server team on how to roll out this change. @lyndaidaii  @jgonz120

lyndaidaii commented:
> We need to have cross team discussion between client and server team on how to roll out this change. @lyndaidaii @jgonz120

I sent out email a while back. I will ping that email thread again.

erdembayar commented:
> > We need to have cross team discussion between client and server team on how to roll out this change. @lyndaidaii @jgonz120
> 
> I sent out email a while back. I will ping that email thread again.

@lyndaidaii 
Do you have any recent update?

srzsantiago commented:
Hi, I see the pull request has been approved but hasn't been merged. Any update on this?

kzu commented:
@joelverhagen is there anything that can be done to avoid the new VS markdown preview (presumably that's the reason?) delaying this any longer?

lyndaidaii commented:
@kzu thank you for your patience. Can you resolve the conflict? I will help verify the change and help to merge to in dev. Meanwhile, we are going to work on the VS readme preview in the future to support the alerts.

kzu commented:
I can do it if the cross-team thing has been resolved. Otherwise, we'll be in the same situation 4 months from now.

lyndaidaii commented:
@kzu, our team decided to support this feature while client team is working on it. if you could resolve the conflict. We could help get this in.

kzu commented:
Rebase done.

Commit:
Merge branch 'dev' into pr-10173-2

Commit:
update unit test

Commit:
update

kzu commented:
Plz do let us know when this hits production on nuget.org so I can start improving my readmes!
